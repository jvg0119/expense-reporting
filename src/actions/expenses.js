import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE   original 
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = 0
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

export const addExpense = (expense) => ({  // modified ADD_EXPENSE 
  type: 'ADD_EXPENSE',
  expense
});

// move the default in startAddExpense 
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });  // end of then()
  };
  // console.log('from startAddExpense ! ')
};

// startAddExpense()

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});


//////////////////////////// 
// test only delete later or use as notes 
/*
const data = {
  a: "real first data",
  b: "real second data"
}
// const {a} =data
// console.log(a)
const makeData = (data = {}) => {
  return () => {
    const {a = 'first default data', b = 'second default data'} = data // default data
    console.log('returning a function inside a function !!!')
    console.log(a)
  }
}
*/
// makeData(data)();







