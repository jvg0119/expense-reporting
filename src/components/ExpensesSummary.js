import React from 'react'; 
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00'); 
  return (
    <div>
      Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
    </div>
  )
}  

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
}


// below is alt method but not setup for testing 
// const ExpensesSummary = (props) => {
//   const expenseWord = props.expenseCount.length === 1 ? "expense" : "expenses";
//   const formattedExpensesTotal = numeral(props.expensesTotal/100).format('$0,0.00')
//   return (
//     <div>
//       Viewing {props.expenseCount.length} {expenseWord} totalling {formattedExpensesTotal}
//     </div>
//   )
// }

// const mapStateToProps = (state, props) => {
//   const visibleExpenses = selectExpenses(state.expenses, state.filters)
//   return {
//     expenseCount: visibleExpenses,
//     expensesTotal: selectExpensesTotal(visibleExpenses)
//   }
// }
  
export default connect(mapStateToProps)(ExpensesSummary);
