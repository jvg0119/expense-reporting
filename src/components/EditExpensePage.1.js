import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

// const EditExpensePage = (props) => {
export const EditExpensePage = (props) => { 
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          // props.dispatch(editExpense(props.expense.id, expense));
          props.editExpense(props.expense.id, expense);
          props.history.push('/');
        }}
      />
      <button onClick={() => {
        // props.dispatch(removeExpense({ id: props.expense.id }));
        props.removeExpense({ id: props.expense.id });
        props.history.push('/');
      }}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, expense) => {
      dispatch(editExpense(id, expense))
    },
    removeExpense: (id) => {
      dispatch(removeExpense(id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// Experiment     --- this works   no test for this experiment 
// using the mapDispatchToProps  but keeping the stateless component 
// not using the class component 
