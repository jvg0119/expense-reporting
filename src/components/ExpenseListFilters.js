import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    // this.props.dispatch(setStartDate(startDate));
    // this.props.dispatch(setEndDate(endDate));
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    // console.log('from onFocusChange >>> ', calendarFocused)
    // below are different versions of setState 
    // this.setState({ calendarFocused })
    // this.setState({ calendarFocused: calendarFocused })
    // this.setState(() => {
    //   return {
    //     calendarFocused: calendarFocused
    //   }
    // })
    // this.setState(() => ({ calendarFocused: calendarFocused }) );
    this.setState(() => ({ calendarFocused }));
  }
  
  onTextChange = (e) => {
    // this.props.dispatch(setTextFilter(e.target.value));
    this.props.setTextFilter(e.target.value);
    // console.log('onTextChange is running !!!')
  }

  onSortBy = (e) => {
    if (e.target.value === 'date') {
      // this.props.dispatch(sortByDate());
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          // onChange={(e) => {
          //   this.props.dispatch(setTextFilter(e.target.value));
          // }}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy}
          // onChange={(e) => {
          //   if (e.target.value === 'date') {
          //     this.props.dispatch(sortByDate());
          //   } else if (e.target.value === 'amount') {
          //     this.props.dispatch(sortByAmount());
          //   }
          // }}
          onChange={this.onSortBy}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

// refactor these from    ExpenseListFilters.js 
//   break out the inline functions 
//   add mapDispatchToProps


