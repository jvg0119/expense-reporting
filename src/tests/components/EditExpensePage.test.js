import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
import moment from '../__mocks__/moment';

let editExpense, removeExpense, historySpy, expense, wrapper, bingo;
beforeEach(() => {
  editExpense = jest.fn(),
  removeExpense = jest.fn(),
  historySpy = {push: jest.fn()}, // changed name from history to historySpy to understand how it uses this var
  expense = expenses[0],
  bingo = 'myBingo',   // you can use these variables in your regular testing 
  wrapper = shallow(
    <EditExpensePage 
      editExpense={editExpense} 
      history={historySpy} // history is the prop being passed; historySpy mock function defined above
      removeExpense={removeExpense} 
      expense={expense} 
    />)
})

test('should render editExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
})

it('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense)
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense)  // you can use expenses[0]
  // these are reqular variables
  // console.error('expense >>> ', expense) 
  // console.error('bingo >>> ', bingo)
  // console.error('>>>> ', historySpy.push)
})

it('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(historySpy.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({id: expense.id}) 
})








