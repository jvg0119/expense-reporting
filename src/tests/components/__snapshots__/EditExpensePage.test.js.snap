// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render editExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 80000,
        "createdAt": 0,
        "description": "Rent",
        "id": "1",
        "note": "this month rent",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
